[ADD] fwbot

antonag32 opened the pull request:
This new module automatically creates forward ports whenever a merge
request is sucessfully merged. It supports Gitlab but a different
provider may be added (in a new model) and used.

antonag32 committed:
[ADD] fwbot

This new module automatically creates forward ports whenever a merge
request is sucessfully merged. It supports Gitlab but a different
provider may be implemented and used.

antonag32 commented:
Hello @moylop260 @luisg123v 

This is the work I have been doing on the bot so far. I am testing it on a dummy repo. For example here is an original MR:

https://git.vauxoo.com/antonag32/opa/-/merge_requests/9

And here is an automatic FW created by the bot:

https://git.vauxoo.com/antonag32/opa/-/merge_requests/10

Instead of doing git operations I am just interacting with the API to create MRs from the original source branches. This makes everything easier but has the disadvantage of branches not being able to be deleted automatically upon merge.

To fix this I came up with two possible solutions so far:

1. Bot modifies elegible MRS and prevents deletion on merge
    - Once request is merged bot creates a new branch based on the merged branch (can be done through the API)
    - It then deletes the original source branch
    - Creates a forward port based on the copied branch
2. Bot keeps track of elegible MRs and keeps a copy branch up to date.
    - Once the branch is merged the bot only has to create a MR with the copy branch as the source
    - Main disadvantage is branch needs to be kept up to date, so that will be extra work (tracking changes and pushing them to copy branch)

I personally like option 1, and think the bot could probably comment something like this when an eligible version is found:

_This MR is eligible for a forward port, the source branch won't be deleted upon merging, forward bot will take care of it_


What do you think?

moylop260 commented on .docker/odoo/Dockerfile:
```diff
@@ -3,7 +3,7 @@ FROM ubuntu:18.04
 RUN apt update; \
     DEBIAN_FRONTEND=noninteractive apt -y --no-install-recommends install \
       ca-certificates curl gnupg lsb-release python3-pip python3-dev python3-lxml libsasl2-dev libldap2-dev ssh \
-      libjpeg-dev libpq-dev libssl1.0-dev build-essential python3-matplotlib nodejs-dev node-gyp npm nginx git locales;  \
```
when you choose `nginx` boolean option from runbot.repo the runbot starts a nginx process

So, I'm not sure if we should remove it

antonag32 replied:
True although I dont think it is setup so if enabled it would not work :thinking:, for instance port 80 is already used by the nginx container itself

moylop260 commented on .docker/odoo/Dockerfile:
```diff
@@ -12,13 +12,10 @@ RUN apt update; \
 
 USER runbot
 RUN mkdir -p ~/instance; \
-    git clone -b 11.0 --depth=1  https://github.com/Vauxoo/runbot-addons.git ~/instance/dependencies/runbot-addons; \
```
I prefer the old option

Since that it could have the whole packages installed and cloned without run extra steps locally

(for devops it is easy)

Even you can develop using vscode + remote container in order to use the source from the container

moylop260 replied:
...even if the repository cloned is the same than this one

We could use directly a `docker push` to quay or docker hub

antonag32 replied:
I think a build arg could be a nice idea to either use a local repo or setup odoo for development, however this is an old project and hopefully one day we won't even develop on it haha so I will revert it to focus on the bot itself

moylop260 commented on forwardbot_gitlab/models/merge_request.py:
```diff
@@ -0,0 +1,214 @@
+            return self.forward_port_id
+
+        try:
+            forward_branch = str(float(self.target_branch) + 1)
```
It is good for me and It will work for 99% of the cases

But we had one time in the past a branch called `saas-x` (not float format)

We could think in the future how to support not-float-format branches

Something like:
`runbot_branch.foward_branch_id = fields.Many2one("runbot.branch")` and maybe auto-choose using this way (float-format) but we can assign another one or skip a middle version of something else.

But for now is good, only add a comment TODO related, please

luisg123v replied:
What happens if we're forward-porting from 14.0 to 16.0?

14.0 + 1 = 15.0.

Please take into account skipping versions is a very common case, most customers don't migrate to the very next version.

antonag32 replied:
> What happens if we're forward-porting from 14.0 to 16.0?
> 
> 14.0 + 1 = 15.0.
> 
> Please take into account skipping versions is a very common case, most customers don't migrate to the very next version.

Good point. I tried to keep this as simple as possible but I think the bot will also need to keep track of the repo branches. Keeping track of this will also stop the bot from making MRs for something like 17.0. In this case the flow will be changed to make a forward port to the next branch (that exists) available

moylop260 replied:
Maybe next sticky branch

But How do you know what is next if you have saas-14 and master

Maybe, the fw_branch_id field could help us a lot here

FYI we have sticky field in the branches to detect the stable ones

luisg123v replied:
For now, I agree on converting branches to float, I just don't agree on adding 1. I think we would need to:
- Convert all branches to float
- Sort them and save the result to a list
- Get index of current branch on the sorted list and add 1 to that index, that should return the next version

How does Odoo this same thing?

Maybe a config param

moylop260 replied:
That could work to auto-assign the fw_branch_id with a value by default

But even we are able to change it from the runbot.branch itself

Consider master, main, saas-* doesn't comply this standard

antonag32 commented:
The bot now tracks protected branches (API does not provide the sticky attribute) and considers those as stable. It sorts them and uses them to get the next version available. It should also support basically all naming conventions, as long as they have a version in the name, or are `main` or `master`. This means it basically works everywhere (in theory).

An initial test run seemed to work fine but I will do more testing.

moylop260 commented on forwardbot_gitlab/api.py:
```diff
@@ -0,0 +1,67 @@
+
+def get_api_data(env):
+    config = env["ir.config_parameter"].sudo()
+    token = config.get_param("forwardbot_gitlab.token")
```
IMHO the token should be configured for each runbot.repo

Similar to current token

Since that each repo could be used from different gitlab server

e.g. gitlab.com, git.vauxoo.com, git.islamic, git.other and so on

Same with base_url, it could be obtained from the runbot.repo name directly

moylop260 replied:
I just realized the runbot.repo doesn't exists here since that this module doesn't depends on runbot modules

So, I need to re-check my last review with this in mind

moylop260 commented on forwardbot_gitlab/models/project.py:
```diff
@@ -0,0 +1,93 @@
+        branches = self.protected_branches.split(",")
+        try:
+            index = branches.index(branch_name)
+            return branches[index + 1]
```
I like this way to configure a default value from runbot.branch to auto-fill a new field similar to

```python
class RunbotBranch(...):
    _inherit = "runbot.branch"

    branch_fw_id =  fields.Many2one("runbot.branch", domain=[("id", "!=", self.id)])

# Create constraint to avoid choosing the same branch than the current one
```

Disclaimer: This code doesn't work, only it is a example to explain the needed change

So, the code should use the value configured in the branch directly in order to be able to change it manually by the user if it is not configuring the correct one branch

moylop260 commented on forwardbot_gitlab/models/project.py:
```diff
@@ -0,0 +1,93 @@
+            try:
+                payload = response.json()
+                branches = [branch["name"] for branch in payload]
+                repo.update_branches(branches)
```
Why is it needed?

antonag32 replied:
It keeps tracks stable branches so that forward ports are correctly created to the next branch

moylop260 commented on forwardbot_gitlab/models/merge_request.py:
```diff
@@ -0,0 +1,276 @@
+
+                # Delete original source branch
+                response = client.delete_branch(merge_request.source_project_id, merge_request.source_branch)
+                if response.status_code not in [204, 404]:
```
response.raise_for_status instead

moylop260 commented on forwardbot_gitlab/models/merge_request.py:
```diff
@@ -0,0 +1,276 @@
+                response = client.create_branch(
+                    forward_port.source_project_id, forward_port.source_branch, merge_request.source_branch
+                )
+                if response.status_code == 400:
```
IMHO we should use response.raise_for_status and process the result of the message of the API response

moylop260 commented on forwardbot_gitlab/models/merge_request.py:
```diff
@@ -0,0 +1,276 @@
+        pruned_mrs = self.browse([])
+        client = ForwardbotClient(url, token, timeout)
+        for merge_request in merge_requests:
+            if merge_request.target_branch in merge_request.target_project_id.protected_branches:
```
I think not all the stable branches for odoo-version are configured as protected_branches

and not all the protected_branches could be related for an odoo-version

Even if it should be protected_branches we should be able to configure the sticky branches similar to runbot.branch

Even if it could be auto-configured using this way

Note: I'm not sure about my own comment

But we need to discuss it in order to have better context

antonag32 replied:
It was the only API way I could think to filter out branches, since I found no reference on the REST API. I visualized this a zero config setup, but I have added an option to overwrite branches and add your own.

moylop260 commented on forwardbot_gitlab/__manifest__.py:
```diff
@@ -0,0 +1,15 @@
+    "category": "Bots",
+    "version": "11.0.1.0.0",
+    "depends": ["base"],
+    "external_dependencies": {"python": ["requests"]},
```
You don't need to add external_dependencies where Odoo already have added:
 - https://github.com/odoo/odoo/blob/01717dc71bae261dad8ad6515ff2b1e2be861295/requirements.txt#L41

```suggestion
```

antonag32 commented:
I have added some options to manually setup the token and URL per repository, falling back to a global ir.config paramter when not available. Stable branches can now be manually setup as well. 

Finally I am working on making this platform independent, each repo has a platform parameter so that the correct method for each provider (github, gitlab) can be used. For the moment only gitlab will be implemented, but adding github or any other provider support should be very easy. Should finish this part soon.

antonag32 commented:
Also changed the name to fwbot since this will no longer be gitlab exclusive. I did not name it forwardbot because Odoo already has a module named like that.

antonag32 commented:
Hey @luisg123v 

I showed the current state of fwbot to @moylop260  and we want to know what you think about it.  Basically the current state is the following:

1. Once an MR is closed, if a valid branch exists (it correctly handles all sort of branch names, at least the ones used) a forward  port is created.
2. To create a forward port, each commit from the MR is cherry-picked into the new MR
3. If one of the cherry-pick fails, the cherrry-picking will stop. All previous cherry-picks that were successful will be kept.
5. If no cherry-pick succeeds, an empty MR is created.

No diff is directly committed when cherry-picking fails, so users need to solve conflicts on a local repository. I don't know how much of a dealbreaker this is or how many conflicts will appear (if they are far in between it may not be that bad).

What  do you think? The biggest plus of this approach is for the developer/maintenance point of view since the code as of now does not require on runbot and does not have to deal with any manual git, so no cloned repos or anything, everything is handled through the API (even cherry-picking).

Of course we need to find a balance between code complexity and user features so please share your thoughts.

antonag32 commented:
Friendly reminder about the pending review @luisg123v :point_right: :point_left:

luisg123v commented on fwbot/models/__init__.py:
```diff
@@ -0,0 +1 @@
+from . import remote_mixin, merge_request, repository  # isort:skip
```
- Place one import per line
- Name files according to classes, e.g. for the model `fwbot.merge.request`, file should be named as `fwbot_merge_request`

moylop260 replied:
Why are you using `isort skip`?

antonag32 replied:
Fixed this, I was running isort without pre-commit-vauxoo so I added that temporarily. I removed it and now run pre-commit-vauxoo just on this module (and not all the other ones in the repo.)

luisg123v replied:
Missing applying 2nd point, files names should match model names

luisg123v commented on fwbot/models/merge_request.py:
```diff
@@ -0,0 +1,313 @@
+class MergeRequest(models.Model):
+    _name = "fwbot.merge.request"
+    _description = "Merge Request (fwbot)"
+    _inherit = "fwbot.remote.mixin"
```
order is: `_name`, `_inherit` and `_description`

luisg123v commented on fwbot/models/merge_request.py:
```diff
@@ -0,0 +1,313 @@
+
+class MergeRequest(models.Model):
+    _name = "fwbot.merge.request"
+    _description = "Merge Request (fwbot)"
```
Merge Request for Fwbot

luisg123v commented on fwbot/models/merge_request.py:
```diff
@@ -0,0 +1,313 @@
+    global_id = fields.Integer(help="ID for this MR on the remote server. Unique across all projects")
+    internal_id = fields.Integer(help="Internal ID. Unique inside the target project.")
+    source_branch = fields.Char(required=True)
+    source_repository_id: Repository = fields.Many2one("fwbot.repository", required=True)
```
This is not a standard Odoo syntax, I would suggest using something standard

luisg123v commented on fwbot/models/repository.py:
```diff
@@ -0,0 +1,155 @@
+    from odoo.api import multi
+except ImportError:
+
+    def multi(func):
```
I don't find this necessary, just use the standard Odoo way.

Let's try to keep code as simple as possible, avoiding hacks.

luisg123v commented on fwbot/models/merge_request.py:
```diff
@@ -0,0 +1,313 @@
+    target_branch = fields.Char(required=True)
+    target_repository_id: Repository = fields.Many2one("fwbot.repository", required=True)
+    state = fields.Selection(
+        [
```
Nitpick:

```suggestion
        selection=[
```

luisg123v commented on fwbot/models/merge_request.py:
```diff
@@ -0,0 +1,313 @@
+
+        forward_branch = self.target_repository_id.get_next_branch(self.target_branch)
+        if not forward_branch:
+            raise ValueError("can't create forward port, missing forward branch")
```
This is a lint

```suggestion
            raise ValueError(_("can't create forward port, missing forward branch"))
```

Same in similar cases.

antonag32 replied:
I think it only applies to Odoo exceptions, I will change it to ValidationError and translate it.

luisg123v commented on fwbot/models/merge_request.py:
```diff
@@ -0,0 +1,313 @@
+            if self.name.startswith("["):
+                name = f"[FW]{name}"
+            else:
+                name = f"[FW] {name}"
```
Use an inline conditional

luisg123v commented on fwbot/models/merge_request.py:
```diff
@@ -0,0 +1,313 @@
+                    "state": data_object["state"],
+                }
+            ),
+            True,
```
What is this for?

I'd suggest using the `key=val` syntax

antonag32 replied:
I added a comment on the return parameters, it indicates whether the record already existed or was created.

luisg123v commented on fwbot/tests/common.py:
```diff
@@ -0,0 +1,86 @@
+from ..models.repository import Repository
+
+
+class MockResponse:
```
Why not using request-mock?

antonag32 replied:
Wanted to keep outside dependencies as less as possible.

luisg123v commented on fwbot/tests/test_merge_request.py:
```diff
@@ -0,0 +1,113 @@
+
+
+class TestMergeRequest(ForwardBotCase):
+    def test_obtain_from_event_create(self):
```
Numerate tests, e.g.:

```suggestion
    def test_01_obtain_from_event_create(self):
```

luisg123v commented on fwbot/tests/test_repository.py:
```diff
@@ -0,0 +1,62 @@
+        self.assertEqual(1000.0, self.model_repo.get_branch_version("main"))
+        self.assertEqual(1000.0, self.model_repo.get_branch_version("master"))
+
+        self.assertEqual(12.0, self.model_repo.get_branch_version("12.0"))
```
Usually it's:

`assertEqual(gotten, expected)`

Same in all other similar cases.

luisg123v commented on fwbot/tests/test_webhook.py:
```diff
@@ -0,0 +1,45 @@
+
+
+class TestWebhook(HttpCase, ForwardBotCase):
+    test_tags = {"standard", "post_install"}
```
Use the standard way

antonag32 replied:
I did this temporarily because the decorator does not exist on Odoo 11, and to speed up development I am running unit tests on Odoo 16, as you can't choose to only run certain tests on Odoo 11, but must run the whole suit AFAIK.

luisg123v replied:
OK, but don't forget to remove it. I'll leave this thread opened until that.

luisg123v commented on fwbot/__manifest__.py:
```diff
@@ -0,0 +1,19 @@
+    "summary": "Automatically create forward ports",
+    "author": "Vauxoo",
+    "website": "https://github.com/Vauxoo/runbot-addons",
+    "license": "AGPL-3",
```
LGPL-3

And update headers accordingly

In fact, remove all headers

luisg123v commented on fwbot/views/fwbot_menu.xml:
```diff
@@ -0,0 +1,5 @@
+<?xml version="1.0" encoding="utf-8" ?>
```
Encoding is preferred to be uppercase, i.e. `UTF-8`

Same in all other XML files.

antonag32 committed:
[FIX] fwbot: lint and style errors

Applied review suggested changes and brought the module up to
pre-commit-vauxoo standard.

antonag32 commented:
I have made the requested changes

luisg123v commented on fwbot/models/merge_request.py:
```diff
@@ -0,0 +1,296 @@
+
+        for merge_request in pending_for_fw:
+            try:
+                merge_request = merge_request.with_context(**{"timeout": timeout})
```
Just do:

```suggestion
                merge_request = merge_request.with_context(timeout=timeout)
```

luisg123v commented on fwbot/models/repository.py:
```diff
@@ -0,0 +1,139 @@
+        for repo in repositories:
+            try:
+                _logger.info("Syncing branches for %s", repo)
+                branches = repo.with_context(**{"timeout": timeout}).get_stable_branches()
```
Same here

luisg123v commented on fwbot/tests/common.py:
```diff
@@ -0,0 +1,83 @@
+            },
+        }
+
+    def assertMergeRequestEquals(self, data: Dict, record):  # pylint: disable=invalid-name
```
- equals -> equal, as the "s" at the end is deprecated for assertion method's names
- Remove types

luisg123v commented on fwbot/tests/test_webhook.py:
```diff
@@ -0,0 +1,45 @@
+    def send_payload(self, payload):
+        return self.url_open("/fwbot_gitlab", data=json.dumps(payload), headers=self.gitlab_headers)
+
+    def test_create_mr(self):
```
Missing numerating tests on this file

antonag32 commented:
Thanks for the review @luisg123v, I think I have covered all comments, just missing the eventual removal of `test_tags` hack. Functionally does it look good to you to go in a beta stage? The main I problem I see with the current approach is that in case of rebase failure, no PR is performed (unlike Odoo which commits the git issues)

luisg123v reviewed (approved):
Looks good, although I think the code is too extensive (maybe it's expected).

moylop260 commented:
@antonag32 What is missing here? (`draft` title)

antonag32 commented:
> @antonag32 What is missing here? (`draft` title)

I wanted to see if I could improve the code but haven't had time, we can merge and start testing if  the current functionality and limitations are fine with you, I will remove the Draft title since we are not waiting on anything else, sorry.

antonag32 commented:
How should we proceed? Should we merge and start looking into deployment? 

cc @moylop260 @luisg123v

desdelinux commented:
Hello @moylop260  and @luisg123v 

This idea needs to become a reality, and I would love for us to start implementing it as soon as possible.

We can begin deploying it with some clients or, ideally, within Vauxoo's own projects.

Put it into action, and improve it as we go.
